### PR TITLE
Parser for hwdetect

### LINKED ISSUE #90: Rewrite `worker::hwdetect::parse_range` using NOM

### PR DESCRIPTION
**Pull request by User3574**
Fixes #90

### CODE REVIEW

**Comment by Kobzol** [comment]
Looks good, but `all_consuming` is a problem, since the input can have whitespace. Either strip the whitespace beforehand or allow whitespace in NOM (the first should be simpler).

**Comment by spirali** [comment]
The code does not cover all cases, newlines and "," delimiters. See tests.

File: src/worker/hwdetect.rs [hunk 1]
```diff
@@ -1,24 +1,44 @@
...
+}
+
+fn p_cpu_ranges(input: &str) -> NomResult<Vec<CpuId>> {
+    separated_list1(tag(","), p_cpu_range)(input)
+        .map(|(a, b)| (a, b.into_iter().flatten().collect()))
+}
+
+pub fn parse_range(input: &str) -> anyhow::Result<Vec<CpuId>> {
+    let f_input = remove_whitespace(input);
+    let res = all_consuming(p_cpu_ranges)(&*f_input)
```
Comment (Kobzol) [inline src/worker/hwdetect.rs:38]: `f_input.as_str()`
and you can also remove the `let res` and just return the expression without `;`

**Review by Kobzol — APPROVED** [review]

File: src/worker/hwdetect.rs [hunk 2]
```diff
@@ -1,24 +1,44 @@
...
-        }
-    }
-    Ok(result)
+use nom::bytes::complete::tag;
+use nom::combinator::all_consuming;
+use nom::combinator::{map_res, opt};
+use nom::multi::separated_list1;
+use nom::sequence::{preceded, tuple};
+
+fn remove_whitespace(s: &str) -> String {
```
Comment (spirali) [inline src/worker/hwdetect.rs:12]: Do not use this kind of method as it blindly removes spaces and makes the combinators uncomposable.
Use space1 from nom. Something like: terminated(xxx, space1) in places where space is allowed

Comment (User3574) [reply src/worker/hwdetect.rs:12]: Updated at my local repo [ https://github.com/User3574/hyperqueue/tree/hwdetect_parser ].

**Pull request merged**